File: geo2jou.py
import os, sys
import re

print(os.getcwd())
FileName = 'pmpol.geo'
cubit.reset()
with  open(FileName, 'r', encoding="utf8") as fid:
  strLines = fid.readlines()

#for n in range(len(strLines)):
for n in range(77):
  strLine = strLines[n]
  if len(strLine)==0:
    print("")
  elif strLine[:2]=='//':
    print("")
  elif strLine[:4]=='Mesh':
    print(strLines[n])
  elif strLine[:5]=='Point':
    pattern = r"Point\((?P<id>\d+)\) = \{(?P<x>[+-]?0|[+-]?\d+\.?\d*(?:[Ee][+-]?\d+)),(?P<y>[+-]?0|[+-]?\d+\.?\d*(?:[Ee][+-]?\d+)),(?P<z>[+-]?0|[+-]?\d+\.?\d*(?:[Ee][+-]?\d+)),hmesh\};"
    match = re.match(pattern, strLine)
    vertex = cubit.create_vertex(float(match.group('x')), float(match.group('y')), float(match.group('z')))
    vertex.set_entity_name('Point_' + match.group('id'))
  elif strLine[:4]=='Line':
    pattern = r"Line\((?P<id>\d+)\) = \{(?P<p1>\d+),(?P<p2>\d+)\};"
    match = re.match(pattern, strLine)
    curve = cubit.create_curve(cubit.vertex(int(match.group('p1'))), cubit.vertex(int(match.group('p2'))))
    curve.set_entity_name('Line_' + match.group('id'))
  elif strLine[:6]=='Circle':
    pattern = r"Circle\((?P<id>\d+)\) = \{(?P<p1>\d+),(?P<p2>\d+),(?P<p3>\d+)\};"
    match = re.match(pattern, strLine)
    cubit.cmd(f'create curve arc center vertex {int(match.group("p2"))} {int(match.group("p1"))} {int(match.group("p3"))}')
    curve.set_entity_name('Circle_' + match.group('id'))
  elif strLine[:10]=='Curve Loop':
    print(strLine[0:20])
#   pattern = r"Curve Loop\((?P<id>\d+)\) = \{(?P<p1>\d+),(?P<p2>\d+),(?P<p3>\d+),(?P<p4>\d+)\};"
    pattern = r"Curve Loop\((?P<id>\d+)\) = \{(?P<pp>.*?)\}(?=(?:\s+|=))"
    match = re.match(pattern, strLine)
    print(match.group('id'))

File: geo2jou.m
clear all;
close all;

FileName = 'pmpol.geo';
strLines = readlines(FileName);

FileName = replace(FileName,'.geo','.jpu');
fid = fopen(FileName,'w');
fprintf(fid,'reset\n');

for n = [1:length(strLines)]
	disp(sprintf('line %d',n));
	if startsWith(strLines(n), 'Point')
		data = sscanf(strLines(n),'Point(%d) = {%e,%e,%e,hmesh};',4);
		fprintf(fid,'create vertex %e %e %e\n',data(2:4));
	end
end
for n = [1:length(strLines)]
	if startsWith(strLines(n), 'Line')
		data = sscanf(strLines(n),'Line(%d) = {%d,%d};',3);
		fprintf(fid,'create curve vertex %d %d\n',data(2:3));
	elseif startsWith(strLines(n), 'Circle')
		data = sscanf(strLines(n),'Circle(%d) = {%d,%d,%d};',4);
		fprintf(fid,'create curve arc center vertex %d %d %d\n',data([3,2,4]));
	end
end
for n = [1:length(strLines)]
	if startsWith(strLines(n), 'Curve Loop')
		id = sscanf(strLines(n),'Curve Loop(%d) = ',1);
		Loop{id} = replace(char(regexp(strLines(n), '{(.*?)}', 'match')), {',','-','{','}'}, {' ','','',''});
	end
end
for n = [1:length(strLines)]
	if startsWith(strLines(n), 'Plane Surface')
		id = sscanf(strLines(n),'Plane Surface(%d) = ',1);
		matches = replace(char(regexp(strLines(n), '{(.*?)}', 'match')), {',','{','}'}, {' ','',''});
		ids = sscanf(matches, '%d');
		if length(ids)==1
			fprintf(fid,'create surface curve %s\n',char(Loop(ids)));
		elseif length(ids)==2
			fprintf(fid,'create surface curve %s\n',char(Loop(ids(1))));
			fprintf(fid,'#{id1=Id("surface")}\n');
			fprintf(fid,'create surface curve %s\n',char(Loop(ids(2))));
			fprintf(fid,'#{id2=Id("surface")}\n');
			fprintf(fid,'subtract surface {id2} from surface {id1} imprint\n');
		end
	end
end
nodesets = 0;
for n = [1:length(strLines)]
	if startsWith(strLines(n), 'Physical Surface')
		nodesets = nodesets + 1;
		name = replace(char(regexp(strLines(n), '("(.*?)")', 'match')), {'(',')','"'}, {'','',''});;
		matches = replace(char(regexp(strLines(n), '{(.*?)}', 'match')), {',','{','}'}, {' ','',''});
		ids = sscanf(matches, '%d');
		nodeset_list = "";
		for m = 1:length(ids)
			id = ids(m);
			nodeset_list = strcat(nodeset_list, char(Loop(id)), " ");
		end
		fprintf(fid,'nodeset %d add curve %s\n',nodesets, nodeset_list);
		fprintf(fid,'nodeset %d name "%s"\n',nodesets, name);
	end
end
fclose(fid);

strLines = readlines(FileName);
clipboard('copy', strjoin(strLines, char(10)));
